cmake/ExampleProject/CMakeLists.txt

```diff
@@ -0,0 +1,20 @@
+cmake_minimum_required(VERSION 3.20)
+project(libCEED_consumer LANGUAGES C CXX)
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_EXPORT_COMPILE_COMMANDS 1)
+
+include(FetchContent)
+
+# libraries that want to use libCEED need only append this in their CMakeLists.txt
+FetchContent_Declare(
+  libCEED
+  GIT_REPOSITORY git@github.com:samuelpmish/libCEED.git
+  GIT_TAG main
+)
+FetchContent_MakeAvailable(libCEED)
```

Comment:
How does this pattern work if one wants other libraries to link the same version? E.g., you want PETSc and MFEM to use the same version of libCEED that you might call directly here?

Reply:
CMake supports a few different options:
- `FetchContent` (shown here) will both download a dependency and build it as part of the current project. 
- `ExternalProject` will download and build (and/or install) a dependency, but keep it separate from the parent project.
- `find_package` will look for an installed dependency in the usual system locations for each platform (or wherever you tell it to look)

So, if I had a project that used libCEED directly and also indirectly through mfem
```
my_project
├── libCEED_version_X.Y.Z
└── mfem
    └── libCEED_version_X.Y.Z
```

 I could either:
1. use the FetchContent to add libCEED to the project and set OVERRIDE_FIND_PACKAGE, so that when mfem uses `find_package(libCEED ...)` it'll pick up the one I requested.
2. use `ExternalProject` to build libCEED as well as mfem, and set `MFEM_CEED_DIR` (or whatever their option is) to the built libCEED
3. have the parent project just write `find_package(libCEED ...)`, and mfem's `find_package(libCEED...)` will find the same one.

When would I want to use choices 1, 2, or 3 above?
1. if I'm frequently modifying libCEED and am interested in seeing how it affects mfem. If they're built as the same project, then I can modify libCEED sources, and those changes will propagate to mfem and the parent project on the next `make`.
2. If I just want to automate dependency resolution for the user, but don't anticipate modifying the dependencies.
3. If I already have a version of libCEED installed and just want to use that, rather than building from source.

These aren't necessarily mutually exclusive either. For instance, one could first check for an installed version (3) and if it can't find something, then do either (1) or (2).

Reply:
In the tree above, `my_project` arguably has global perspective. But what if the dependence on `libCEED` is in a sibling library to `mfem` (thus not aware of `mfem` or `my_project`. Which of the above options can just work when `my_project` includes `mfem` and that sibling library?

Reply:
if you have
```
my_project
├── foo
|   └── libbaz_version_X.Y.Z
└── bar
    └── libbaz_version_X.Y.Z
```
foo and bar can both check if `find_package(baz)` succeeds, and fallback on FetchContent (for example) if it fails. However, if they don't have that logic in their CMake project, then the parent project needs to be responsible for setting up its subproject's dependencies.

That being said, CMake isn't a package manager. Although it can manage dependencies, for big projects it's often not the right tool for that job. However, if a project has a reasonable CMake implementation, integration with actual package managers like vcpkg and conan is easier.